### PR TITLE
Improving approve transaction flow

### PR DESCRIPTION
**Pull request by jpuri**
Improve approve transaction flow

https://app.zenhub.com/workspaces/confirmations-6245e6e2348677001213b8d2/issues/gh/metamask/metamask-planning/317

### CODE REVIEW

File: app/scripts/controllers/transactions/lib/util.js [hunk 1]
```diff
@@ -127,10 +128,21 @@ function ensureProperTransactionEnvelopeTypeProvided(txParams, field) {
  * Validates the given tx parameters
  *
  * @param {object} txParams - the tx params
+ * @param {string} type - type of the transaction
```
Comment (digiwand) [inline app/scripts/controllers/transactions/lib/util.js:131]: I think we'll want to update `normalizeAndValidateTxParams` to support the new param as well

File: ui/components/app/transaction-decoding/transaction-decoding.component.js [hunk 2]
```diff
@@ -99,6 +108,7 @@ export default function TransactionDecoding({ to = '', inputData: data = '' }) {
         }
       }
     })();
+    return () => setMounted(false);
   }, [t, from, to, network, data]);
 
   // ***********************************************************
```
Comment (jpuri) [inline ui/components/app/transaction-decoding/transaction-decoding.component.js:114]: This file's change are not directly related to task,  but were needed to remove warnings from new test cases added.

File: ui/pages/confirm-transaction/confirm-transaction.nomock.test.js [hunk 3]
```diff
@@ -0,0 +1,218 @@
...
+    expect(getAllByText('Approve')).toHaveLength(2);
+    act(() => {
+      const hexTabButton = getByRole('button', { name: 'Hex' });
+      fireEvent.click(hexTabButton);
+    });
+    expect(
+      getByText(sendWithApproveTransaction.txParams.data),
+    ).toBeInTheDocument();
+  });
+});
```
Comment (jpuri) [inline ui/pages/confirm-transaction/confirm-transaction.nomock.test.js:218]: I named this file as `confirm-transaction.nomock.test.js` since @digiwand is also working on test coverage of same file with mocking. https://github.com/MetaMask/metamask-extension/pull/17473

File: ui/pages/confirm-transaction/confirm-transaction.component.js [hunk 4]
```diff
@@ -159,7 +159,12 @@ export default class ConfirmTransaction extends Component {
       transactionId &&
       (!paramsTransactionId || paramsTransactionId === transactionId);
 
-    if (isTokenMethodAction && validTransactionId) {
+    const value = transaction?.txParams?.value;
+    if (
+      isTokenMethodAction &&
+      validTransactionId &&
+      (!value || value === '0x0')
```
Comment (digiwand) [inline ui/pages/confirm-transaction/confirm-transaction.component.js:166]: I am wondering, why check for `(!value || value === '0x0')` rather than `isSendWithApproval` here?

Comment (jpuri) [reply ui/pages/confirm-transaction/confirm-transaction.component.js]: Updated

File: ui/pages/confirm-transaction/confirm-transaction.component.js [hunk 5]
```diff
@@ -159,7 +160,16 @@ export default class ConfirmTransaction extends Component {
...
 
-    if (isTokenMethodAction && validTransactionId) {
+    let isSendWithApprove = false;
+    if (transaction?.txParams) {
+      const {
+        type,
+        txParams: { value },
+      } = transaction;
+      isSendWithApprove =
+        type === TransactionType.tokenMethodApprove && value && value !== '0x0';
```
Comment (digiwand) [inline ui/pages/confirm-transaction/confirm-transaction.component.js:170]: [suggestion/non-blocker] I think `isSendWithApprove` would be good to add to `ui/helpers/utils/transactions.util.js`

**Review by digiwand — DISMISSED** [review]
👍

**Review by segun — DISMISSED** [review]

File: ui/pages/confirm-transaction/confirm-transaction.component.js (hunk 5 above)
Comment (jpuri) [reply ui/pages/confirm-transaction/confirm-transaction.component.js]: 👍 done

**Comment by jpuri** [comment]
@digiwand , @segun : can you plz re-approve, I updated the PR to resolve conflicts and implement review feedbacks.

@seaona : The PR also needs some QA attention.

**Review by digiwand — DISMISSED** [review]

**Review by digiwand — DISMISSED** [review]

**Comment by seaona** [comment]
I am not 100% confident on reviewing the code at this level but somehow I believe we could re-use some methods/screen that are implemented for the generic Contract Interaction case. With that I mean, taking the example of the Deposit function from the test-dapp, we call a contract function with value. In this case, the exact same result should happen (screen is the same with all the components), so maybe we could re-use some of that? Let me know if this makes sense to you

https://user-images.githubusercontent.com/54408225/217873114-8fa45b51-a08b-4dfc-805c-af9a29c35dce.mp4

**Review by segun — DISMISSED** [review]

**Comment by jpuri** [comment]
Hey @seaona : I fixed recipient address, can you plz retest.

**Comment by seaona** [comment]
I see the issue fixed @jpuri ! Nice work!! I just have a remaining question, I directed in slack and after this is clarified the PR should be ready

**Comment by jpuri** [comment]
Closing in favour of https://github.com/MetaMask/metamask-extension/pull/17777
